cmds/echo/echo.go

```diff
@@ -6,30 +6,174 @@
+type flags struct {
+	noNewline, interpretEscapes bool
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
 
-func echo(w io.Writer, s ...string) error {
-	_, err := fmt.Fprintf(w, "%s", strings.Join(s, " "))
+func isHex(x rune) bool {
+	return (x >= '0' && x <= '9') || (x >= 'A' && x <= 'F')
+}
+
+func isOctal(x rune) bool {
+	return x >= '0' && x <= '7'
+}
 
-	if !*nonewline {
-		fmt.Fprint(w, "\n")
+func parseEncodedCharacter(s string, index int, base int) (rune, int, error) {
+	var (
+		maxRange    int
+		test        func(rune) bool
+		escapedText string
+	)
+	switch base {
+	default:
+		return ' ', 0, fmt.Errorf("Attempting to interpret unsupported base %d\n", base)
+	case 8:
+		maxRange = 3
+		test = isOctal
+	case 16:
+		maxRange = 2
+		test = isHex
 	}
 
+	l := len(s)
+	for _, c := range s[min(index+1, l):min(index+maxRange+1, l)] {
+		if !test(c) {
+			break
+		}
+		escapedText += string(c)
+	}
+
+	l = len(escapedText)
+	if l == 0 {
+		return ' ', 0, fmt.Errorf("No encoded text\n")
+	}
+
+	char, err := strconv.ParseInt(escapedText, base, 8)
+	return rune(char), index + l, err
+}
+
+func escapeString(f *flags, s string) (string, error) {
+	var (
+		buf  bytes.Buffer
+		err  error
+		char rune
+	)
+	l := len(s) - 1
+StringLoop:
+	for i := 0; i < len(s); i++ {
+		char = rune(s[i])
+		if char == '\\' {
+			i++
+			if min(l, i) == l {
+				char = rune(92)
+			} else {
+				switch s[i] {
+				default:
+					char = rune(92)
+					i--
+				case '\\':
+					char = rune(92)
+				case 'a':
+					char = rune(7)
+				case 'b':
+					char = rune(8)
+				case 'c':
+					f.noNewline = true
+					break StringLoop
+				case 'e':
+					char = rune(27)
+				case 'f':
+					char = rune(12)
+				case 'n':
+					char = rune(10)
+				case 't':
+					char = rune(9)
+				case 'v':
+					char = rune(11)
+				case 'x':
+					char, i, err = parseEncodedCharacter(s, i, 16)
+				case '0':
+					char, i, err = parseEncodedCharacter(s, i, 8)
+				}
+				if err != nil {
+					return "", err
+				}
+			}
+		}
+		buf.Write([]byte(string(char)))
+	}
+	return buf.String(), err
+}
+
+func echo(f flags, w io.Writer, s ...string) error {
+	var err error
+	line := strings.Join(s, " ")
+	if f.interpretEscapes {
+		line, err = escapeString(&f, line)
+		if err != nil {
+			return err
+		}
+	}
+
+	format := "%s"
+	if !f.noNewline {
+		format += "\n"
+	}
+	_, err = fmt.Fprintf(w, format, line)
+
 	return err
 }
 
+func init() {
+	defUsage := flag.Usage
+	flag.Usage = func() {
+		defUsage()
+		fmt.Println(`
+  If -e is in effect, the following sequences are recognized:
+    \\     backslash
+    \a     alert (BEL)
+    \b     backspace
+    \c     produce no further output
+    \e     escape
+    \f     form feed
+    \n     new line
+    \r     carriage return
+    \t     horizontal tab
+    \v     vertical tab
+    \0NNN  byte with octal value NNN (1 to 3 digits)
+    \xHH   byte with hexadecimal value HH (1 to 2 digits)`)
+	}
+}
+
 func main() {
+	var f flags
```

Comment:
instead of embedding your own parser for these cases can't you just use the one in fmt? And why not just always interpret the argument as a fmt-style string and skip the option?

Reply:
Hm, yea, I was struggling to see how to do it but finally worked out that `fmt.Sscanf(s, "%q", &s)` will interpret the escapes.
As for the flag, I based the usage on gnu `echo` since the roadmap mentioned the `-e` flag.  We should really implement `-E` as well and can make `-e` the default which appears to be the case in `zsh`.